[brein62] iP

brein62 opened the pull request:
# Bob the chatbot
The best **Duke** chatbot in Singapore, JB, *and some say Batam*

Bob the Chatbot allows you to store tasks so that you don't forget them. It's:

- Simple
- Easy to use
- Text-based
- ~Many~ Very few commands to learn!

All you need to do is to...
1. Download ```duke.jar``` from [heree](https://github.com/brein62/ip/releases/tag/A-Jar).
2. Run ```java -jar (file).jar```
3. Add your tasks.
4. **PROFIT** :money_mouth_face: :money_mouth_face: :money_mouth_face:

> "I don't know why, but this is a very nice app to use" - me

Features:

- [ ] Managing different tasks (todo, reminders, events)
- [ ] Storing and saving tasks
- [X] Mark tasks as completed!

If you're a Java programmer, you can use it to practice Java too. Here's the `main` method to store the tasks in a file called `./data/duke.txt`:
```java
public static void main(String[] args) {
    Path path = Paths.get("data", "duke.txt");
    new Duke(path).run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

brein62 committed:
Add skeleton for Level 0

brein62 committed:
Add echo function for Level 1

brein62 committed:
Add list functionality for Level 2

brein62 committed:
Add Task class and add functionality to mark tasks as done for Level 3

brein62 committed:
Add different task types for Level 4

brein62 committed:
Perform automated text UI testing for Level 4

brein62 committed:
Add DukeException and add most error cases such that they throw DukeException for Level 5

brein62 committed:
Add delete functionality and handle exceptions for mark, unmark for Level 6

brein62 committed:
Separate long strings into separate lines

brein62 committed:
Add save functionality for Level 7

brein62 committed:
Add save functionality for Level 7 (code files)

brein62 committed:
Add dates functionality using LocalDateTime for Level 8

brein62 committed:
Add more classes to improve OOP in app

brein62 committed:
Refactor and add classes to packages

Add files (except for main Duke class) into the duke.* package,
including two subpackages: duke.command and duke.task for command and
task-related classes respectively.

brein62 committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2324S1/ip

brein62 committed:
Add Gradle functionality

Change the main class to Duke and add a sample dummy test function for JUnit testing

brein62 committed:
Add JUnit test cases and fix date formatting bug

brein62 committed:
Fix Text UI Test to reflect date bugfix

brein62 committed:
Add JavaDoc for all Command classes

brein62 committed:
Move Duke class into duke package

brein62 committed:
Fix style for Command classes and change to ui.print()

brein62 committed:
Fix style for duke and duke.task classes

brein62 committed:
Add JavaDoc for duke and duke.task classes

brein62 committed:
Add find functionality for Level 9

brein62 committed:
Add find functionality and test cases for Level 9

brein62 committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java

brein62 committed:
Merge branch 'branch-Level-9'

brein62 committed:
Consolidate changes from A-JavaDoc and Level-9

brein62 committed:
Add Checkstyle XML files

brein62 committed:
Modify files to adhere to checkstyle

brein62 committed:
Modify files to adhere to checkstyle

brein62 committed:
Implement JavaFX GUI

Add a simple JavaFX GUI inspired by the Duke tutorial's chatbot layout.
Removed most text UI related elements from project and re-did logic to
use the new GUI instead of the text UI (commands executed now return a
String instead of directly getting the UI to print them).

brein62 committed:
Edit JavaDoc to remove text UI references and refactor MainWindow

tanveersingh10 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,88 @@
+     * @throws DukeException If there is an invalid input message causing an error in parsing.
+     */
+    public static Command parse(String message) throws DukeException {
+        if (message.equals("bye")) {
```
I think this method could look a lot cleaner with the switch / case syntax. You could break it up into smaller methods as well

brein62 replied:
thank you for the feedback! didn't realise that switch-case syntax can also be used for strings as well so will make the necessary changes 👍 . i agree that this parse method could be broken up into multiple methods since the ```parse``` function implementation is very long

tanveersingh10 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,127 @@
+     *
+     * @param tasks The list of tasks to be saved into the file.
+     */
+    public void saveTasks(TaskList tasks) {
```
I like the variable naming. Does exactly what you'd expect.

tanveersingh10 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,137 @@
+ */
+public class AddCommand extends Command {
+
+    enum CommandType {
```
good use of enums! something i didnt do lol

tanveersingh10 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,42 @@
+     *
+     * @return The deadline, formatted as a String to be saved in the save file.
+     */
+    public String toSaveFormatString() {
```
very clear method naming! not too long and not too short

tanveersingh10 reviewed (commented):
LGTM!

Darren159 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,65 @@
+package duke;
+
+import java.nio.file.Path;
+
+import duke.command.Command;
+import duke.gui.GuiApplication;
+import duke.task.TaskList;
+import javafx.application.Application;
```
I think making the ordering of import statements would help enhance the readability of the code.

brein62 replied:
thank you for the feedback! actually `import` statements aren't exactly the most well-defined in the CS2103T coding standards, so i just followed the order recommended by checkstyle:

- static imports
- `java.*` imports, followed by
- other imports (including `duke`/`javafx`)

but will check the files again and see if there are any coding standard violations for `import`s i may have missed out! 👍

Darren159 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,127 @@
+        Task newTask;
+        String[] lineSeq = line.split(" \\| ");
+        try {
+            boolean isValidCompletion = (lineSeq.length < 2)
+                ? false
+                : (Objects.equals(lineSeq[1], "1") || Objects.equals(lineSeq[1], "0"));
+
+            boolean isValidTodo = (line.startsWith("T") && lineSeq.length == 3);
+            boolean isValidDeadline = (line.startsWith("D") && lineSeq.length == 4);
+            boolean isValidEvent = (line.startsWith("E") && lineSeq.length == 5);
```
I like that your naming of booleans adheres to java coding standards!

Darren159 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,137 @@
+     *         "Add Event" if AddCommand adds an event; "Add " otherwise.
+     */
+    @Override
+    public String getCommandType() {
+        String typeStr;
+        switch (this.type) {
+        case TODO:
+            typeStr = "Todo";
+            break;
+        case DEADLINE:
+            typeStr = "Deadline";
+            break;
+        case EVENT:
+            typeStr = "Event";
+            break;
+        default:
+            typeStr = "";
+            break;
+        }
+        return "Add " + typeStr;
+    }
+}
```
I like that your switch statement follows java coding standards!

Darren159 reviewed (commented):
Overall I like that your variable and method naming follows java coding standards and there is extensive use of Javadocs. LGTM!

brein62 committed:
Change Parser to adhere to code quality guideline

The Parser.java file contains a parse() method that parses the commands
input by the user into the chatbot. However, it is a very long method
and could be split up.

The if-else syntax is replaced by a switch-case syntax with minor
modifications to the logic to accommodate this syntax (should not be
noticed by users with normal use). This improves code readability.

Additionally, most multi-statement code is replaced by more private
static methods within the Parser class to ease readability of the
parse() method even more.

This fix in code quality was inspired and spearheaded by the comments
in this PR: https://github.com/nus-cs2103-AY2324S1/ip/pull/288

brein62 committed:
Refactor LocalDateTime formatting

Formatting for LocalDateTime is expected to be consistent for the entire
program when displayed to the user in the GUI.

By introducing a new Formatter utility class with a single static method
formatDateTime, it reduces the amount of clutter within the methods
in which they are used. This also ensures consistency in the formatting
of LocalDateTime as String objects within the GUI.

In addition, removed unnecessary this.* references from methods when
referring to instance fields.

brein62 committed:
Add JUnit test cases for duke.util.Formatter

brein62 committed:
Merge pull request #1 from brein62/branch-A-CodeQuality

brein62 committed:
Add assertions to Parser and MainWindow

To ensure that certain private methods will only be run in certain
scenarios, assertions are used to enforce a precondition at the start
of

Parser class: Ensure that the different commands parsed, as handled by
different private methods within the Parser class only happen when the
message starts with the corresponding command message. For example,
only inputs starting with 'mark' can trigger the parseMarkCommand()
private static method in Parser and return a MarkCommand object.

MainWindow class: Ensure that the handleExit() method is only run if the
user specifies and types "bye". Checks if the input message starts with
"bye".

Gradle is also configured to handle assertions in this commit.

brein62 committed:
Merge pull request #2 from brein62/branch-A-Assertions

brein62 committed:
Add stream functionality to TaskList

To improve performance to the TaskList, the pure ArrayList
implementation of TaskList has been changed to a more generic java.util.
List implementation, which also allows for more functional code through
Streams.

Within the TaskList, the getSavedStrings() and find() methods have been
modified to use stream implementation. Additionally, a new constructor
for TaskList based on Stream<Task> has been created.

Other changes:
- Change test files and other classes to fit the new implementation of
  TaskList that doesn't solely rely on ArrayList anymore
- Change other files to adhere to Checkstyle

brein62 committed:
Merge pull request #3 from brein62/branch-A-Streams

brein62 committed:
Add gradle.yml file for Github CI

brein62 committed:
Add update and clone functionality

The ability to update and clone tasks without deleting them or adding
the details all over again helps the user to improve the ease of use of
the application.

By adding the update functionality, tasks can now be updated using
simple commands such as `update 4 msg Hello World` or `update 3 Deadline
2023-10-10T12:34:56`. This updates the message and deadline for the
tasks respectively. This is done by making the Tasks more mutable and
allowing for updates to be directly done to the task.

By adding the clone functionality, users can add Tasks based on another
task without having to specify the task all over again. This may be
helpful for recurring tasks or related tasks.

These implementations are for the extension **C-Update**.

brein62 committed:
Add JUnit test cases for Update, Clone and Find

brein62 committed:
Merge pull request #4 from brein62/branch-C-Update

brein62 committed:
Improve colour appearance in GUI

The chatbot so far does not distinguish between error responses by the
bot and regular responses. When there is an error response, a dark red
reply will be shown by the bot. Dark red is chosen to improve contrast
and thus readability of the chatbot.

The background colours are changed to improve distinction between the
bot and the human input. The bot has a default colour palette (same as
background) while the human input has a silver colour palette.

brein62 committed:
Fix imports to adhere to Checkstyle

brein62 committed:
Merge pull request #6 from brein62/branch-A-BetterGui

brein62 committed:
Add detection to ensure Event start is before end

With the previous implementation, Event end date/time can be set to be
before the start date/time. With this detection in place, when creating
or updating an Event, there is a check in place to ensure Event from
date/time is before Event to date/time. This check is done in the
constructor, when updating using the `update()` method, as well as when
cloning the Event.

brein62 committed:
Improve error handling when parsing Events

Previously, when the parser parses an event command, the parser will
disregard the logic of /from and /to markers within the event command.

This fixes the issue and ensures that exactly 1 /from and 1 /to marker
is present in the event command and a few JUnit test cases under
parseEvent have been added.

JUnit test cases:
- 2 /from: DukeException thrown
- 2 /to: DukeException thrown
- /from and /to flipped: accepted and event is flipped as needed

brein62 committed:
Merge pull request #7 from brein62/branch-A-MoreErrorHandling

brein62 committed:
Add test cases for update methods in Task subclasses

brein62 committed:
Add JUnit test cases for individual task clone methods

brein62 committed:
Merge pull request #8 from brein62/branch-A-MoreTesting

brein62 committed:
Add title, test data and program screenshot

brein62 committed:
Update README.md in user guide folder

brein62 committed:
Update docs README.md table due to GitHub pages bug

brein62 committed:
Update docs README.md table due to GitHub pages bug

brein62 committed:
Update docs README.md table due to GitHub pages bug

brein62 committed:
Remove docs README.md table

brein62 committed:
Remove extra character from README.md

brein62 committed:
Fix bug where data/dukeGui.txt is not created as intended

brein62 committed:
Merge pull request #9 from brein62/fix-data-folder-bug

brein62 committed:
Add product title to the front of README.md
